File: .github/scripts/build-index.js
//scripts/build-index.js
import fs from 'node:fs';
import path from 'node:path';

const ROOT = process.cwd();
const BLOG_DIR = path.join(ROOT, 'blog');
const OLD_DIR  = path.join(ROOT, 'old_posts');

const ensureDir = p => fs.existsSync(p) || fs.mkdirSync(p, { recursive: true });
const writeJSON = (p, obj) => fs.writeFileSync(p, JSON.stringify(obj, null, 2) + '\n');
const readText  = p => fs.existsSync(p) ? fs.readFileSync(p, 'utf8') : '';

function findMD(base) {
    const out = [];
    if (!fs.existsSync(base)) return out;
    (function walk(dir){
        for (const name of fs.readdirSync(dir)) {
            const full = path.join(dir, name);
            const st = fs.statSync(full);
            if (st.isDirectory()) walk(full);
            else if (st.isFile() && name.toLowerCase().endsWith('.md')) out.push(full);
        }
    })(base);
    return out;
}

function guessTitle(file, md) {
    const m1 = md.match(/(?:^|\n)\s*title:\s*["']?(.+?)["']?\s*$/mi);
    if (m1) return m1[1].trim();
    const m2 = md.match(/^(#|##)\s+(.+)$/m);
    if (m2) return m2[2].trim();
    return path.basename(file).replace(/\.md$/i, '');
}

function guessDate(file) {
    const b = path.basename(file);
    const m = b.match(/^(\d{2})(\d{2})(\d{4})/); // DDMMYYYY
    return m ? m[0] : '00000000';
}

function parseDateParts(ddmmyyyy) {
    return { dd: ddmmyyyy.slice(0,2), mm: ddmmyyyy.slice(2,4), yyyy: ddmmyyyy.slice(4,8) };
}

//move older posts into /old_posts/YYYY/MM
function archiveOldPosts() {
    ensureDir(BLOG_DIR);
    ensureDir(OLD_DIR);

    const files = findMD(BLOG_DIR).filter(f => path.dirname(f) === BLOG_DIR);
    const posts = files.map(f => {
        const md = readText(f);
        return {
            abs: f,
            path: '/' + path.relative(ROOT, f).replace(/\\/g,'/'),
            date: guessDate(f),
            title: guessTitle(f, md)
        };
    }).sort((a,b) => b.date.localeCompare(a.date));

    const keep = posts.slice(0, 10);
    const archive = posts.slice(10);

    for (const p of archive) {
        const { mm, yyyy } = parseDateParts(p.date);
        const destDir = path.join(OLD_DIR, yyyy, mm);
        ensureDir(destDir);
        const destPath = path.join(destDir, path.basename(p.abs));

        //move file
        fs.renameSync(p.abs, destPath);
        console.log(`Archived: ${p.abs} → ${destPath}`);
    }
}

//rebuild blog index (latest 10)
function buildBlogIndex() {
    ensureDir(BLOG_DIR);
    const files = findMD(BLOG_DIR).filter(f => path.dirname(f) === BLOG_DIR);
    const items = files.map(f => {
        const md = readText(f);
        return {
            path: '/' + path.relative(ROOT, f).replace(/\\/g,'/'),
            date: guessDate(f),
            title: guessTitle(f, md)
        };
    }).sort((a,b) => b.date.localeCompare(a.date));
    writeJSON(path.join(BLOG_DIR, 'index.json'), items);
}

//rebuild archive indexes
function buildArchiveIndexes() {
    if (!fs.existsSync(OLD_DIR)) return;
    const mdFiles = findMD(OLD_DIR); // /old_posts/YYYY/MM/*.md

    const years = new Set();
    const byYear = new Map();

    for (const f of mdFiles) {
        const rel = path.relative(ROOT, f).replace(/\\/g,'/');
        const parts = rel.split('/');
        if (parts.length < 4) continue;
        const y = parts[1], mm = parts[2];
        years.add(y);
        const md = readText(f);
        const entry = { path: '/' + rel, date: guessDate(f), title: guessTitle(f, md) };
        if (!byYear.has(y)) byYear.set(y, []);
        byYear.get(y).push(entry);
    }

    const yearsList = Array.from(years).sort();
    writeJSON(path.join(OLD_DIR, 'index.json'), yearsList);

    for (const y of yearsList) {
        const items = (byYear.get(y) || []).sort((a,b)=> b.date.localeCompare(a.date));
        const yearDir = path.join(OLD_DIR, y);
        ensureDir(yearDir);
        writeJSON(path.join(yearDir, 'index.json'), items);

        const byMonth = new Map();
        for (const it of items) {
            const mm = it.path.split('/')[3];
            if (!byMonth.has(mm)) byMonth.set(mm, []);
            byMonth.get(mm).push(it);
        }
        for (const [mm, arr] of byMonth) {
            const monthDir = path.join(OLD_DIR, y, mm);
            ensureDir(monthDir);
            writeJSON(path.join(monthDir, 'index.json'), arr.sort((a,b)=> b.date.localeCompare(a.date)));
        }
    }
}

// --- Run all ---
archiveOldPosts();
buildBlogIndex();
buildArchiveIndexes();

console.log('Indexes rebuilt with archiving.');

File: .github/scripts/build-index.py
#!/usr/bin/env python3
"""
Fast index rebuilds:
- Always rebuilds blog/index.json (10 newest in /blog)
- By default, only rebuilds the most recent archive month: old_posts/<latestYYYY>/<latestMM>/index.json
- Set env FULL_REINDEX=1 to rebuild ALL old_posts/**/index.json
"""
import os, re, json, pathlib

ROOT = pathlib.Path(os.getcwd())
BLOG_DIR = ROOT / "blog"
OLD_DIR  = ROOT / "old_posts"

def ensure_dir(p: pathlib.Path): p.mkdir(parents=True, exist_ok=True)

def write_json(p: pathlib.Path, obj):
    p.parent.mkdir(parents=True, exist_ok=True)
    with p.open("w", encoding="utf-8") as f:
        json.dump(obj, f, indent=2, ensure_ascii=False)
        f.write("\n")

def read_title_from_md(path: pathlib.Path) -> str:
    try:
        with path.open("r", encoding="utf-8") as f:
            for line in f:
                if line.startswith("# "):
                    return line[2:].strip()
    except FileNotFoundError:
        pass
    return ""

def collect_posts(base_dir: pathlib.Path):
    posts = []
    if not base_dir.exists():
        return posts
    for name in os.listdir(base_dir):
        if re.match(r"^\d{8}-.*\.md$", name):
            dmy = name[:8]
            dd, mm, yyyy = dmy[0:2], dmy[2:4], dmy[4:8]
            ymd = f"{yyyy}-{mm}-{dd}"
            full = base_dir / name
            posts.append({
                "date": ymd,
                "file": str(full.relative_to(ROOT)).replace("\\", "/"),
                "title": read_title_from_md(full) or name[9:-3]
            })
    posts.sort(key=lambda x: x["date"], reverse=True)
    return posts

def build_blog_index():
    ensure_dir(BLOG_DIR)
    posts = collect_posts(BLOG_DIR)
    write_json(BLOG_DIR / "index.json", posts)

def build_archive_index_for(month_dir: pathlib.Path):
    posts = collect_posts(month_dir)
    write_json(month_dir / "index.json", posts)

def find_latest_archive_month():
    if not OLD_DIR.exists():
        return None
    years = [d for d in OLD_DIR.iterdir() if d.is_dir() and d.name.isdigit()]
    if not years:
        return None
    latest_year = max(years, key=lambda p: int(p.name))
    months = [d for d in latest_year.iterdir() if d.is_dir() and d.name.isdigit()]
    if not months:
        return None
    latest_month = max(months, key=lambda p: int(p.name))
    return latest_month

def build_archive_indexes(full: bool = False):
    if not OLD_DIR.exists():
        return
    if full:
        for yyyy in sorted([d for d in OLD_DIR.iterdir() if d.is_dir() and d.name.isdigit()], key=lambda p: int(p.name)):
            for mm in sorted([d for d in yyyy.iterdir() if d.is_dir() and d.name.isdigit()], key=lambda p: int(p.name)):
                build_archive_index_for(mm)
        return
    latest = find_latest_archive_month()
    if latest:
        build_archive_index_for(latest)

if __name__ == "__main__":
    try:
        import subprocess, sys
        subprocess.run([sys.executable, str(ROOT / ".github" / "scripts_py" / "archive_posts.py")], check=False)
    except Exception:
        pass

    build_blog_index()
    full = os.environ.get("FULL_REINDEX", "").strip() in ("1", "true", "True", "yes")
    build_archive_indexes(full=full)
    mode = "full" if full else "fast"
    print(f"Indexes rebuilt ({mode} mode).")
